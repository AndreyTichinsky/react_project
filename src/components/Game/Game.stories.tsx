import React from "react";
import { withKnobs, number, boolean, text } from "@storybook/addon-knobs";

import { Game } from "./Game";

export default {
  title: "Game",
  decorators: [withKnobs],
};

export const instanceGame = () => (
  <Game
    xSize={number("xSize", 10)}
    ySize={number("ySize", 10)}
    cellSize={number("cellSize", 50)}
    updateSpeed={text("updateSpeed", "slow")}
    gameInProgress={boolean("gameInProgress", false)}
    nameIsSubmited={boolean("nameIsSubmited", false)}
  />
);
