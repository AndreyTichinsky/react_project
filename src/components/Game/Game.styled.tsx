import styled from "@emotion/styled";
/** @jsx jsx */
import { css, jsx } from "@emotion/core";

interface GameWrapperStyle {
  xSize: number;
  cellSize: number;
}

export const GameWrapper = styled.div`
  min-height: 500px;
  background-color: #fff;
  position: relative;
  margin: 50px auto 0;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  min-width: ${(props: GameWrapperStyle) => props.xSize * props.cellSize}px;
`;
