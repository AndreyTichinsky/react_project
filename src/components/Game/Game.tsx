import React from "react";

type FieldComponentInterface = React.FC<{
  field: boolean[][];
  cellSize: number;
  onClick: (x: number, y: number) => void;
}>;

interface GameProps {
  xSize: number;
  ySize: number;
  cellSize: number;
  fieldComponent: FieldComponentInterface;
}
interface GameState {
  fieldState: boolean[][];
}

export class Game extends React.Component<GameProps, GameState> {
  private FieldComponent: FieldComponentInterface;
  constructor(props: GameProps) {
    super(props);
    this.FieldComponent = props.fieldComponent;
    this.state = {
      fieldState: Array.from({ length: props.ySize }).map(() =>
        Array.from({ length: props.xSize }).fill(false)
      ) as boolean[][],
    };
    this.onClick = this.onClick.bind(this);
  }
  public onClick(x: number, y: number) {
    this.setState((state) => {
      const fieldStateCopy = state.fieldState.map((row) => [...row]);
      fieldStateCopy[y][x] = !fieldStateCopy[y][x];
      return {
        fieldState: fieldStateCopy,
      };
    });
  }
  render() {
    const FieldComponent = this.FieldComponent;
    return (
      <FieldComponent
        field={this.state.fieldState}
        onClick={this.onClick}
        cellSize={this.props.cellSize}
      />
    );
  }
}
