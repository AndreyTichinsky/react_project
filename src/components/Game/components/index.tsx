export * from "./Field";
export * from "./Menu";
