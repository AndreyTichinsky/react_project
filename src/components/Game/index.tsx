export * from "./Game";
export * from "./components";
