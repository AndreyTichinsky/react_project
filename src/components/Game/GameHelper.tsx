import type { BooleanMatrix } from "types/game";
import { Cache } from "./Game";

export const makeMatrix = (
  y: number,
  x: number,
  filledPercent: number
): BooleanMatrix => {
  if (filledPercent === 0) {
    return makeEmptyMatrix(y, x);
  }
  return makeFilledMatrix(y, x, filledPercent);
};

export const makeFilledMatrix = (
  y: number,
  x: number,
  filledPercent: number
): BooleanMatrix => {
  const cellsAmount = x * y;
  const filledCellsAmount = Math.round((filledPercent * cellsAmount) / 100);
  const arrayOfIdxs: number[][] = [];
  const ratio = filledCellsAmount / cellsAmount;
  const isFilled = ratio >= 50;
  const rawArray = Array.from({ length: y }).map((row, i) => {
    return Array.from({ length: x }).map((cell, j) => {
      arrayOfIdxs.push([i, j]);
      return isFilled;
    });
  });
  const arrayOfFilledCells = shuffle(arrayOfIdxs, filledCellsAmount);
  arrayOfFilledCells.forEach(([i, j], index) => {
    rawArray[i][j] = !rawArray[i][j];
  });

  return rawArray;
};

export const shuffle = (arr: number[][], filledCellsAmount: number) => {
  const arrayCopy = arr.slice();
  let currentIdx = arrayCopy.length,
    randomIdx;

  while (currentIdx !== 0) {
    randomIdx = Math.floor(Math.random() * currentIdx);
    currentIdx--;

    [arrayCopy[currentIdx], arrayCopy[randomIdx]] = [
      arrayCopy[randomIdx],
      arrayCopy[currentIdx],
    ];
  }

  return arrayCopy.slice(0, filledCellsAmount);
};

export const makeEmptyMatrix = (y: number, x: number): BooleanMatrix => {
  return Array.from({ length: y }).map(() => {
    return Array.from({ length: x }).map(() => {
      return false;
    });
  });
};

export const mergeMatrices = (
  oldMatrix: BooleanMatrix,
  newMatrix: BooleanMatrix
): BooleanMatrix => {
  const aLenX = oldMatrix[0].length,
    aLenY = oldMatrix.length,
    bLenX = newMatrix[0].length,
    bLenY = newMatrix.length,
    copyA = oldMatrix.map((row) => [...row]),
    copyB = newMatrix.map((row) => [...row]);
  let mainArr: BooleanMatrix = copyA;
  if (aLenY >= bLenY) {
    copyA.splice(bLenY);
  } else {
    mainArr = copyB;
  }
  if (aLenX >= bLenX) {
    mainArr = mainArr.map((row, i) => {
      if (copyA[i] && copyB[i]) {
        return copyA[i].slice(0, bLenX);
      }
      return mainArr[i];
    });
  } else {
    mainArr = mainArr.map((row, i) => {
      if (copyA[i] && copyB[i]) {
        return copyA[i].slice(0, aLenX).concat(copyB[i].slice(aLenX));
      }
      return mainArr[i];
    });
  }
  return mainArr;
};

export const matrixSum = (matrix: BooleanMatrix): number => {
  return matrix.reduce((mainAcc, row) => {
    return (
      mainAcc +
      row.reduce((subAcc, el) => {
        return subAcc + Number(el);
      }, 0)
    );
  }, 0);
};

export const calculatePercentage = (matrix: BooleanMatrix): number => {
  return Math.floor(
    (matrixSum(matrix) / (matrix.length * matrix[0].length)) * 100
  );
};

export const isNumber = (item: number): boolean => {
  return !Number.isNaN(item);
};

export const assertSizeValue = (value: number): boolean => {
  return isNumber(value) && value > 0;
};

export const assertZero = (value: number | null): boolean => {
  return value !== null && value !== 0;
};

export const assertPercentValue = (value: number): boolean => {
  return value >= 0 && value <= 100;
};

export const generationGenerator = (
  matrix: BooleanMatrix,
  cache: Cache
): BooleanMatrix => {
  const nextGenMatrix: BooleanMatrix = matrix.map((row, i) => {
    return row.map((cell, j) => {
      let counter = 0;
      cache[`${i}_${j}`].forEach(([y, x]) => {
        counter += matrix[y][x] ? 1 : 0;
      });

      return (
        (cell && (counter === 2 || counter === 3)) || (!cell && counter === 3)
      );
    });
  });
  return nextGenMatrix;
};

export const cacheNeighbours = (matrix: BooleanMatrix) => {
  const cache: Cache = {};
  const matrixYLen = matrix.length;
  const matrixXLen = matrix[0].length;
  matrix.forEach((row, i) => {
    row.forEach((cell, j) => {
      cache[`${i}_${j}`] = getNeighbours(j, i, matrixXLen, matrixYLen);
    });
  });
  return cache;
};

export const getNeighbours = (
  x: number,
  y: number,
  xLen: number,
  yLen: number
) => {
  const pattern = [
    [y - 1, x - 1],
    [y - 1, x],
    [y - 1, x + 1],
    [y, x - 1],
    [y, x + 1],
    [y + 1, x - 1],
    [y + 1, x],
    [y + 1, x + 1],
  ];
  return pattern.map(([Y, X]) => {
    if (Y >= xLen) Y %= xLen;
    if (X >= yLen) X %= yLen;
    if (Y < 0) Y = xLen + Y;
    if (X < 0) X = yLen + X;
    return [Y, X];
  });
};
