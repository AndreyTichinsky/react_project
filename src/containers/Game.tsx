import { connect } from "react-redux";
import { Game } from "@/components/Game/Game";
import { State } from "@/redux/reducers";

const mapStateToProps = (state: State) => ({
  xSize: state.xSize,
  ySize: state.ySize,
  cellSize: state.cellSize,
  initialPercent: state.initialPercent,
  updateSpeed: state.updateSpeed,
  gameInProgress: state.gameInProgress,
  fieldState: state.fieldState,
});

export default connect(mapStateToProps)(Game);
