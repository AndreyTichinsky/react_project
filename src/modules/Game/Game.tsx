import React, { FC } from "react";
import { Field, Menu, LogoutForm, StartForm } from "@/modules";
import type { FieldData } from "types/index";
import { GameWrapper } from "./Game.styled";
import { AppDispatch, State } from "@/redux";
import { actions } from "./reducer";
import { connect } from "react-redux";

const mapStateToProps = (state: State) => ({
  xSize: state.game.xSize,
  ySize: state.game.ySize,
  cellSize: state.game.cellSize,
  initialPercent: state.game.initialPercent,
  updateSpeed: state.game.updateSpeed,
  gameInProgress: state.game.gameInProgress,
  fieldState: state.game.fieldState,
});

interface GameProps {
  xSize: number;
  ySize: number;
  updateSpeed: string;
  initialPercent: number;
  gameInProgress: boolean;
  cellSize: number;
  fieldState: FieldData;
  dispatch: AppDispatch;
  onLogout: (ev: React.FormEvent) => void;
  [index: string]: any;
}

const GameComponent: FC<GameProps> = (props) => (
  <GameWrapper cellSize={props.cellSize} xSize={props.xSize}>
    <LogoutForm onLogout={props.onLogout} />
    <Menu />
    <Field />
    <StartForm
      actionSelect={actions.setSpeed.type}
      actionProgress={actions.setProgress.type}
    />
  </GameWrapper>
);

export const Game = connect(mapStateToProps)(GameComponent);
