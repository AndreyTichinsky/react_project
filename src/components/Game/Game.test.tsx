import React from "react";
import { mount, shallow } from "enzyme";
import { Game } from "./Game";
import { calculatePercentage } from "./GameHelper";

const mock = jest.fn();

const defaultProps = {
  xSize: 2,
  ySize: 2,
  cellSize: 50,
  updateSpeed: "slow",
  gameInProgress: false,
  username: "Guest",
  onLogout: mock,
};

const formEvent = {
  preventDefault: (): void => {
    void 0;
  },
} as React.FormEvent;

describe("Game", () => {
  it("when init game expect props equality to defaultProps", () => {
    const wrapper = mount(<Game {...defaultProps} />);
    expect(wrapper.props()).toEqual({
      xSize: 2,
      ySize: 2,
      cellSize: 50,
      updateSpeed: "slow",
      gameInProgress: false,
      username: "Guest",
      onLogout: mock,
    });
  });

  it("when set new props expect equality to certain props", () => {
    const wrapper = mount(<Game {...defaultProps} />);
    wrapper.setProps({
      xSize: 5,
      ySize: 5,
    });
    expect(wrapper.props()).toEqual({
      xSize: 5,
      ySize: 5,
      cellSize: 50,
      updateSpeed: "slow",
      gameInProgress: false,
      username: "Guest",
      onLogout: mock,
    });
  });

  it("when set gameInProgress property to true expect progress button to change text to 'stop'", () => {
    const wrapper = mount(<Game {...defaultProps} />);
    wrapper.setProps({
      gameInProgress: true,
    });
    expect(wrapper.find("button.progress_button").text()).toEqual("Stop");
  });

  it("when simulate click to progress button expect gameInProgress to be truthy", () => {
    const wrapper = mount(<Game {...defaultProps} />);
    wrapper.find("button.progress_button").simulate("click");
    expect(wrapper.state("gameInProgress")).toBeTruthy();
  });

  it("when speed select change to value 'fast' expect updateSpeed to equals 'fast'", () => {
    const wrapper = mount(<Game {...defaultProps} />);
    wrapper
      .find("select.speed_select")
      .simulate("change", { target: { value: "fast" } });
    expect(wrapper.state("updateSpeed")).toEqual("fast");
  });

  // handlers testing
  it("when invoke handleReset expect fieldState to be empty", () => {
    const wrapper = shallow<Game>(<Game {...defaultProps} />);
    wrapper.setState({
      fieldState: [
        [true, true],
        [true, true],
      ],
    });
    wrapper.instance().handleReset(formEvent);
    expect(wrapper.state("fieldState")).toStrictEqual([
      [false, false],
      [false, false],
    ]);
  });

  it("when invoke handleUpdate expect fieldState update its size", () => {
    const wrapper = shallow<Game>(<Game {...defaultProps} />);
    expect(wrapper.state("xSize")).toBe(2);
    expect(wrapper.state("ySize")).toBe(2);
    wrapper.setState({
      xSize: 4,
      ySize: 4,
    });
    wrapper.instance().handleUpdate();
    expect(wrapper.state("fieldState")).toStrictEqual([
      [false, false, false, false],
      [false, false, false, false],
      [false, false, false, false],
      [false, false, false, false],
    ]);
    expect(wrapper.state("xSize")).toEqual(4);
    expect(wrapper.state("ySize")).toEqual(4);
  });

  it("when invoke handleProgress expect gameInProgress to be truthy", () => {
    const wrapper = shallow<Game>(<Game {...defaultProps} />);
    expect(wrapper.state("gameInProgress")).toBeFalsy();
    wrapper.instance().handleProgress(formEvent);
    expect(wrapper.state("gameInProgress")).toBeTruthy();
  });

  it("when invoke handleGenerator expect generated matrix to equals state properties", () => {
    const wrapper = shallow<Game>(<Game {...defaultProps} />);
    expect(wrapper.state("xSize")).toBe(2);
    expect(wrapper.state("ySize")).toBe(2);
    expect(wrapper.state("initialPercent")).toBe(0);
    wrapper.setState({
      xSize: 4,
      ySize: 4,
      initialPercent: 50,
    });
    wrapper.instance().handleGenerator(formEvent);
    const testMatrix = wrapper.state("fieldState");
    const percent = calculatePercentage(testMatrix);
    expect(percent).toEqual(50);
    expect(testMatrix.length).toEqual(4);
    expect(testMatrix[0].length).toEqual(4);
  });
  it("when invoke handleFilledPercent expect initialPercent state to equals event target value", () => {
    const wrapper = shallow<Game>(<Game {...defaultProps} />);
    expect(wrapper.state("initialPercent")).toEqual(0);
    const mockEvent = {
      target: {
        name: "filledPercent",
        value: "50",
      },
    } as React.ChangeEvent<HTMLInputElement>;
    wrapper.instance().handleFilledPercent(mockEvent);
    expect(wrapper.state("initialPercent")).toEqual(50);
  });
  it("when invoke handleXSizeChange expect xSize state to equals event target value", () => {
    const wrapper = shallow<Game>(<Game {...defaultProps} />);
    expect(wrapper.state("xSize")).toEqual(2);
    const mockEvent = {
      target: {
        name: "xSize",
        value: "4",
      },
    } as React.ChangeEvent<HTMLInputElement>;
    wrapper.instance().handleSize(mockEvent, "handleXSizeChange");
    expect(wrapper.state("xSize")).toEqual(4);
  });
  it("when invoke handleYSizeChange expect ySize state to equals event target value", () => {
    const wrapper = shallow<Game>(<Game {...defaultProps} />);
    expect(wrapper.state("ySize")).toEqual(2);
    const mockEvent = {
      target: {
        name: "ySize",
        value: "4",
      },
    } as React.ChangeEvent<HTMLInputElement>;
    wrapper.instance().handleSize(mockEvent, "handleYSizeChange");
    expect(wrapper.state("ySize")).toEqual(4);
  });
});
