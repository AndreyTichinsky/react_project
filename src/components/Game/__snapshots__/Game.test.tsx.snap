// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game game snapshot test before submit username 1`] = `
.emotion-1 {
  min-height: 500px;
  background-color: #fff;
  position: relative;
  margin: 50px auto 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 100px;
}

.emotion-0 {
  width: 100px;
  margin-right: 10px;
}

<div
  className="emotion-1"
>
  <fieldset
    disabled={true}
  >
    <form>
      <button
        className="progress_button emotion-0"
        onClick={[Function]}
      >
        Start
      </button>
      <label>
        Speed:
        <select
          onChange={[Function]}
          value="slow"
        >
          <option
            value="slow"
          >
            slow
          </option>
          <option
            value="normal"
          >
            normal
          </option>
          <option
            value="fast"
          >
            fast
          </option>
        </select>
      </label>
    </form>
  </fieldset>
</div>
`;
